<a href="https://colab.research.google.com/github/srijita1406/Cross-validation-in-ML/blob/main/Python_Assignment_Quant_ArthAlpha.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import requests
import io
import numpy as np
import pandas as pd

In [ ]:
link = 'https://drive.google.com/file/d/1L33SrY3LK0BeDLJzVVy4LNI-X4nvPhrb/view'
file_id = link.split('d/')[1]
file_id1 = file_id.split('/')[0]
url = f"https://drive.google.com/uc?id={file_id1}&export=download"
response = requests.get(url)

with open('nse500.csv', 'wb') as f:
    f.write(response.content)

In [ ]:
df = pd.read_csv('nse500.csv')
df.head(100)

20020630     20020731     20020831     20020930     20021031  \
0   1909315D IN  1909315D IN  1909315D IN  1909315D IN  1909315D IN   
1   1918689D IN  1918689D IN  1918689D IN  1918689D IN  1918689D IN   
2    294932Q IN   294932Q IN   294932Q IN   294932Q IN   294932Q IN   
3         3M IN        3M IN        3M IN        3M IN        3M IN   
4       ABAN IN      ABAN IN      ABAN IN      ABAN IN      ABAN IN   
..          ...          ...          ...          ...          ...   
95      CUBK IN        CU IN     CSTRL IN    CRISIL IN    CRISIL IN   
96      CYBT IN      CUBK IN        CU IN     CRPBK IN     CRPBK IN   
97       CYL IN      CYBT IN      CUBK IN      CRSA IN      CRSA IN   
98     DABUR IN       CYL IN      CYBT IN     CSTRL IN     CSTRL IN   
99       DCB IN     DABUR IN       CYL IN        CU IN        CU IN   

       20021130     20021231     20030131     20030228     20030331  ...  \
0   1909315D IN  1909315D IN  1909315D IN  1909315D IN  1909315D IN  ...   
1   1918689D IN  1918689D IN  1918689D IN  1918689D IN  1918689D IN  ...   
2    294932Q IN   294932Q IN   294932Q IN   294932Q IN   294932Q IN  ...   
3         3M IN        3M IN        3M IN        3M IN        3M IN  ...   
4       ABAN IN      ABAN IN      ABAN IN      ABAN IN      ABAN IN  ...   
..          ...          ...          ...          ...          ...  ...   
95     CRPBK IN     CRPBK IN     CRPBK IN     CRPBK IN    CRISIL IN  ...   
96      CRSA IN      CRSA IN      CRSA IN      CRSA IN     CRPBK IN  ...   
97     CSTRL IN     CSTRL IN     CSTRL IN     CSTRL IN      CRSA IN  ...   
98        CU IN        CU IN        CU IN        CU IN     CSTRL IN  ...   
99      CUBK IN      CUBK IN      CUBK IN      CUBK IN        CU IN  ...   

       20230331     20230430     20230531     20230630     20230731  \
0     360ONE IS    360ONE IN    360ONE IN    360ONE IN    360ONE IN   
1         3M IS        3M IS        3M IS        3M IS        3M IS   
2       AACL IS      AACL IS      AACL IS      AACL IS      AACL IS   
3      AAVAS IS     AAVAS IS     AAVAS IS     AAVAS IS     AAVAS IS   
4        ABB IS       ABB IS       ABB IS       ABB IS       ABB IS   
..          ...          ...          ...          ...          ...   
95   CGPOWER IS    CHALET IS    CHALET IS    CHALET IS    CHALET IS   
96    CHALET IS  CHEMPLAS IS  CHEMPLAS IS  CHEMPLAS IS  CHEMPLAS IS   
97  CHEMPLAS IS      CHMB IS      CHMB IS      CHMB IS      CHMB IS   
98      CHMB IS  CHOLAHLD IS  CHOLAHLD IS  CHOLAHLD IS  CHOLAHLD IS   
99  CHOLAHLD IS  CIEINDIA IS  CIEINDIA IS  CIEINDIA IS  CIEINDIA IS   

       20230831     20230930     20231031     20231130     20231231  
0     360ONE IN    360ONE IN    360ONE IN    360ONE IN    360ONE IN  
1         3M IS        3M IS        3M IS        3M IS        3M IS  
2       AACL IS      AACL IS      AACL IS      AACL IS      AACL IS  
3      AAVAS IS     AAVAS IS     AAVAS IS     AAVAS IS     AAVAS IS  
4        ABB IS       ABB IS       ABB IS       ABB IS       ABB IS  
..          ...          ...          ...          ...          ...  
95    CHALET IS   CGPOWER IS   CGPOWER IS   CGPOWER IS   CGPOWER IS  
96  CHEMPLAS IS    CHALET IS    CHALET IS    CHALET IS    CHALET IS  
97      CHMB IS  CHEMPLAS IS  CHEMPLAS IS  CHEMPLAS IS  CHEMPLAS IS  
98  CHOLAHLD IS      CHMB IS      CHMB IS      CHMB IS      CHMB IS  
99  CIEINDIA IS  CHOLAHLD IS  CHOLAHLD IS  CHOLAHLD IS  CHOLAHLD IS  

[100 rows x 259 columns]

In [ ]:
df.shape

(502, 259)

Q1.

*   Read the file into the Pandas data frame
*   Stock tickers have either “IN”, “IS”, or “IB” as suffixes with space between the ticker
and suffix. Remove the suffixes and the space from the tickers

Ans: We are using the re module is Python's built-in library which allows for pattern-based string manipulation
we are defining a function so that we can later use the python map function and apply it to the entire dataframe.
We could have done it in parts but this method is faster and more efficient.
Isintance is being used because it was throwing not a string error otherwise so to makesure we used Isinstance.

In [ ]:
import re
# Defining a function to remove the suffixes and spaces from the tickers
def remove_suffix(string):
    if isinstance(string, str):
        # Remove suffixes and extra spaces
        string = re.sub(r' (IN|IS|IB)$', '', string).replace(" ", "")
    return string


# Applying the function to the entire DataFrame and handling not string objects
df = df.apply(lambda col: col.map(remove_suffix) if col.dtype == 'object' else col)
df.head(100)

20020630  20020731  20020831  20020930  20021031  20021130  20021231  \
0   1909315D  1909315D  1909315D  1909315D  1909315D  1909315D  1909315D   
1   1918689D  1918689D  1918689D  1918689D  1918689D  1918689D  1918689D   
2    294932Q   294932Q   294932Q   294932Q   294932Q   294932Q   294932Q   
3         3M        3M        3M        3M        3M        3M        3M   
4       ABAN      ABAN      ABAN      ABAN      ABAN      ABAN      ABAN   
..       ...       ...       ...       ...       ...       ...       ...   
95      CUBK        CU     CSTRL    CRISIL    CRISIL     CRPBK     CRPBK   
96      CYBT      CUBK        CU     CRPBK     CRPBK      CRSA      CRSA   
97       CYL      CYBT      CUBK      CRSA      CRSA     CSTRL     CSTRL   
98     DABUR       CYL      CYBT     CSTRL     CSTRL        CU        CU   
99       DCB     DABUR       CYL        CU        CU      CUBK      CUBK   

    20030131  20030228  20030331  ...  20230331  20230430  20230531  20230630  \
0   1909315D  1909315D  1909315D  ...    360ONE    360ONE    360ONE    360ONE   
1   1918689D  1918689D  1918689D  ...        3M        3M        3M        3M   
2    294932Q   294932Q   294932Q  ...      AACL      AACL      AACL      AACL   
3         3M        3M        3M  ...     AAVAS     AAVAS     AAVAS     AAVAS   
4       ABAN      ABAN      ABAN  ...       ABB       ABB       ABB       ABB   
..       ...       ...       ...  ...       ...       ...       ...       ...   
95     CRPBK     CRPBK    CRISIL  ...   CGPOWER    CHALET    CHALET    CHALET   
96      CRSA      CRSA     CRPBK  ...    CHALET  CHEMPLAS  CHEMPLAS  CHEMPLAS   
97     CSTRL     CSTRL      CRSA  ...  CHEMPLAS      CHMB      CHMB      CHMB   
98        CU        CU     CSTRL  ...      CHMB  CHOLAHLD  CHOLAHLD  CHOLAHLD   
99      CUBK      CUBK        CU  ...  CHOLAHLD  CIEINDIA  CIEINDIA  CIEINDIA   

    20230731  20230831  20230930  20231031  20231130  20231231  
0     360ONE    360ONE    360ONE    360ONE    360ONE    360ONE  
1         3M        3M        3M        3M        3M        3M  
2       AACL      AACL      AACL      AACL      AACL      AACL  
3      AAVAS     AAVAS     AAVAS     AAVAS     AAVAS     AAVAS  
4        ABB       ABB       ABB       ABB       ABB       ABB  
..       ...       ...       ...       ...       ...       ...  
95    CHALET    CHALET   CGPOWER   CGPOWER   CGPOWER   CGPOWER  
96  CHEMPLAS  CHEMPLAS    CHALET    CHALET    CHALET    CHALET  
97      CHMB      CHMB  CHEMPLAS  CHEMPLAS  CHEMPLAS  CHEMPLAS  
98  CHOLAHLD  CHOLAHLD      CHMB      CHMB      CHMB      CHMB  
99  CIEINDIA  CIEINDIA  CHOLAHLD  CHOLAHLD  CHOLAHLD  CHOLAHLD  

[100 rows x 259 columns]

In [ ]:
df.shape

(502, 259)

Q2. • Construct the following data frame from the above data frame:


1.   Unique Tickers as Column Names
2.   Dates as Row indices


Ans:
1.   first we use df.to_numpy to write the values of the dataframe as a numpy array
2.   we then flatten that array into a 1D array and keep only the unique values using pd.unique
3.   we create a new dataframe where we define the list containing the unique values as column headers and the column headers of the previous dataframe as the index






In [ ]:
# Creating set of Unique Tickers as Column Names
unique_tickers = pd.unique(df.to_numpy().flatten())
print(len(unique_tickers))
# Initializing a new DataFrame with Dates as Row indices
result_df = pd.DataFrame(index=df.columns, columns=unique_tickers)
result_df.head()

1215


1909315D 1958850D ABAN   3M ABCAP 360ONE 1918689D 294908Q 2299371D  \
20020630      NaN      NaN  NaN  NaN   NaN    NaN      NaN     NaN      NaN   
20020731      NaN      NaN  NaN  NaN   NaN    NaN      NaN     NaN      NaN   
20020831      NaN      NaN  NaN  NaN   NaN    NaN      NaN     NaN      NaN   
20020930      NaN      NaN  NaN  NaN   NaN    NaN      NaN     NaN      NaN   
20021031      NaN      NaN  NaN  NaN   NaN    NaN      NaN     NaN      NaN   

         20ML  ... WESTLIFE ZYDUSLIF WDJL   ZC  ZDC ZNDU ZUARIIND ZEEN ZCVCS  \
20020630  NaN  ...      NaN      NaN  NaN  NaN  NaN  NaN      NaN  NaN   NaN   
20020731  NaN  ...      NaN      NaN  NaN  NaN  NaN  NaN      NaN  NaN   NaN   
20020831  NaN  ...      NaN      NaN  NaN  NaN  NaN  NaN      NaN  NaN   NaN   
20020930  NaN  ...      NaN      NaN  NaN  NaN  NaN  NaN      NaN  NaN   NaN   
20021031  NaN  ...      NaN      NaN  NaN  NaN  NaN  NaN      NaN  NaN   NaN   

         ZOMATO  
20020630    NaN  
20020731    NaN  
20020831    NaN  
20020930    NaN  
20021031    NaN  

[5 rows x 1215 columns]

Q2.

   part **II**)  
    For each date and ticker combination set the value of the data frame field to 1 if the
   ticker was part of the NSE 500 index on that date; otherwise set the value to “nan”.
   Please use Numpy NaN for the same.


Ans
1. In the solution below we first iterate over each date in the dataFrame where the  dates are columns the tickers are in the result_df . For each date and ticker combination, it checks whether the ticker is present in the values of the corresponding date column in df. If the ticker is found, the corresponding cell in result_df is set to 1; otherwise, it is set to NaN. This process fills result_df with 1 or NaN based on whether each ticker was part of the NSE 500 index on the given date.



a copy was created to not disturb the original result_df

In [ ]:
updated_result_df=result_df.copy()

In [ ]:
import time
start_time = time.time()
for date in df.columns:
    for ticker in result_df.columns:
        if ticker in df[date].values:
            updated_result_df.loc[date, ticker] = 1
        else:
            updated_result_df.loc[date, ticker] = np.nan
updated_result_df.head()
end_time = time.time()
execution_time = end_time - start_time

In [ ]:
updated_result_df.head()

1909315D 1958850D ABAN 3M ABCAP 360ONE 1918689D 294908Q 2299371D  \
20020630        1      NaN    1  1   NaN    NaN        1     NaN      NaN   
20020731        1      NaN    1  1   NaN    NaN        1     NaN      NaN   
20020831        1      NaN    1  1   NaN    NaN        1     NaN      NaN   
20020930        1      NaN    1  1   NaN    NaN        1     NaN      NaN   
20021031        1      NaN    1  1   NaN    NaN        1     NaN      NaN   

         20ML  ... WESTLIFE ZYDUSLIF WDJL   ZC  ZDC ZNDU ZUARIIND ZEEN ZCVCS  \
20020630  NaN  ...      NaN        1    1  NaN  NaN    1        1  NaN   NaN   
20020731  NaN  ...      NaN        1    1  NaN  NaN    1        1  NaN   NaN   
20020831  NaN  ...      NaN        1    1  NaN  NaN    1        1  NaN   NaN   
20020930  NaN  ...      NaN        1  NaN    1    1    1        1  NaN   NaN   
20021031  NaN  ...      NaN        1  NaN    1    1    1        1  NaN   NaN   

         ZOMATO  
20020630    NaN  
20020731    NaN  
20020831    NaN  
20020930    NaN  
20021031    NaN  

[5 rows x 1215 columns]

Q3. If the solution above involves 2 “loops”, refine the solution such that it uses a maximum of 1 loop.

Since the solution above to fill with 1 if it was present on that date had two loops and the time it was taking is greater we reduced it to a single loop

In [ ]:
updated_result_df1=result_df.copy()

In [ ]:
start_time = time.time()
for date in df.columns:
    tickers_on_date =pd.unique(df[date].dropna().values)
    updated_result_df1.loc[date, tickers_on_date] = 1
updated_result_df1.head()
end_time = time.time()
execution_time1 = end_time - start_time

In [ ]:
updated_result_df1.head()

1909315D 1958850D ABAN 3M ABCAP 360ONE 1918689D 294908Q 2299371D  \
20020630        1      NaN    1  1   NaN    NaN        1     NaN      NaN   
20020731        1      NaN    1  1   NaN    NaN        1     NaN      NaN   
20020831        1      NaN    1  1   NaN    NaN        1     NaN      NaN   
20020930        1      NaN    1  1   NaN    NaN        1     NaN      NaN   
20021031        1      NaN    1  1   NaN    NaN        1     NaN      NaN   

         20ML  ... WESTLIFE ZYDUSLIF WDJL   ZC  ZDC ZNDU ZUARIIND ZEEN ZCVCS  \
20020630  NaN  ...      NaN        1    1  NaN  NaN    1        1  NaN   NaN   
20020731  NaN  ...      NaN        1    1  NaN  NaN    1        1  NaN   NaN   
20020831  NaN  ...      NaN        1    1  NaN  NaN    1        1  NaN   NaN   
20020930  NaN  ...      NaN        1  NaN    1    1    1        1  NaN   NaN   
20021031  NaN  ...      NaN        1  NaN    1    1    1        1  NaN   NaN   

         ZOMATO  
20020630    NaN  
20020731    NaN  
20020831    NaN  
20020930    NaN  
20021031    NaN  

[5 rows x 1215 columns]

Question 4\
• If the solution above involves any “loop”, refine the solution such that it uses no loops.

I have implemented two methods here but the second method has better time optimisation.


Ans.\
Method 1:\
In the first method we use a function called update_result function which is designed to populate updated_result_df2 by marking which tickers were part of the index on each specific date. For each date (column) in df, it finds the tickers present and then sets the corresponding ticker cell values in updated_result_df2 to 1.

In [ ]:
updated_result_df2=result_df.copy()

In [ ]:
start_time=time.time()
def update_result(date_series):
    tickers_on_date = pd.unique((date_series).dropna())
    updated_result_df2.loc[date_series.name, tickers_on_date] = 1
df.apply(update_result)
print(updated_result_df2)
end_time = time.time()
execution_time2 = end_time - start_time

         1909315D 1958850D ABAN 3M ABCAP 360ONE 1918689D 294908Q 2299371D  \
20020630        1      NaN    1  1   NaN    NaN        1     NaN      NaN   
20020731        1      NaN    1  1   NaN    NaN        1     NaN      NaN   
20020831        1      NaN    1  1   NaN    NaN        1     NaN      NaN   
20020930        1      NaN    1  1   NaN    NaN        1     NaN      NaN   
20021031        1      NaN    1  1   NaN    NaN        1     NaN      NaN   
...           ...      ...  ... ..   ...    ...      ...     ...      ...   
20230831      NaN      NaN  NaN  1     1      1      NaN     NaN      NaN   
20230930      NaN      NaN  NaN  1     1      1      NaN     NaN      NaN   
20231031      NaN      NaN  NaN  1     1      1      NaN     NaN      NaN   
20231130      NaN      NaN  NaN  1     1      1      NaN     NaN      NaN   
20231231      NaN      NaN  NaN  1     1      1      NaN     NaN      NaN   

         20ML  ... WESTLIFE ZYDUSLIF WDJL   ZC  ZDC ZNDU ZUARIIND ZEEN ZCVC

Ans.\
Method 2.\
In this code we first melt the dataframe into a long format with columns date and stock. We then assign value 1 to the melted_df dataframe tp the date stock combinations and then we use the pivot function to pivot it back to look like the result_df where data are indices and the stocks are the tickers. Thus using pivot we reshape the dataFrame from a long format to a wide format. This method takes less time as can be seen below where we compare execution times.





In [ ]:
updated_result_df3=result_df.copy()

In [ ]:
start_time = time.time()
melted_df = df.melt(var_name='Date', value_name='Stock')
updated_result_df3 = (melted_df.assign(value=1)
                     .pivot_table(index='Date', columns='Stock', values='value'))
stocks = pd.unique(df.to_numpy().flatten())
dates = df.columns
updated_result_df3= updated_result_df3.reindex(index=dates, columns=stocks, fill_value=np.nan)
print(updated_result_df3)
end_time = time.time()
execution_time3 = end_time - start_time

Stock     1909315D  1958850D  ABAN   3M  ABCAP  360ONE  1918689D  294908Q  \
20020630       1.0       NaN   1.0  1.0    NaN     NaN       1.0      NaN   
20020731       1.0       NaN   1.0  1.0    NaN     NaN       1.0      NaN   
20020831       1.0       NaN   1.0  1.0    NaN     NaN       1.0      NaN   
20020930       1.0       NaN   1.0  1.0    NaN     NaN       1.0      NaN   
20021031       1.0       NaN   1.0  1.0    NaN     NaN       1.0      NaN   
...            ...       ...   ...  ...    ...     ...       ...      ...   
20230831       NaN       NaN   NaN  1.0    1.0     1.0       NaN      NaN   
20230930       NaN       NaN   NaN  1.0    1.0     1.0       NaN      NaN   
20231031       NaN       NaN   NaN  1.0    1.0     1.0       NaN      NaN   
20231130       NaN       NaN   NaN  1.0    1.0     1.0       NaN      NaN   
20231231       NaN       NaN   NaN  1.0    1.0     1.0       NaN      NaN   

Stock     2299371D  20ML  ...  WESTLIFE  ZYDUSLIF  WDJL   ZC  ZDC  ZNDU  \


Comparing execution times

In [ ]:
print(f"Execution time 1: {execution_time} seconds")
print(f"Execution time 2: {execution_time1} seconds")
print(f"Execution time 3: {execution_time2} seconds")
print(f"Execution time 4: {execution_time3} seconds")

Execution time 1: 42.6417555809021 seconds
Execution time 2: 0.49918603897094727 seconds
Execution time 3: 0.5048825740814209 seconds
Execution time 4: 0.14576983451843262 seconds


To check homogeinity of the results of the different methods

In [ ]:
column_wise_one_counts_df = (updated_result_df == 1.0).sum(axis=1)
column_wise_one_counts_df1 = (updated_result_df1 == 1.0).sum(axis=1)
column_wise_one_counts_df2 = (updated_result_df2 == 1.0).sum(axis=1)
column_wise_one_counts_df3 = (updated_result_df3 == 1.0).sum(axis=1)

print(column_wise_one_counts_df)
print(column_wise_one_counts_df1)
print(column_wise_one_counts_df2)
print(column_wise_one_counts_df3)

20020630    500
20020731    500
20020831    500
20020930    500
20021031    500
           ... 
20230831    502
20230930    501
20231031    501
20231130    501
20231231    501
Length: 259, dtype: int64
20020630    500
20020731    500
20020831    500
20020930    500
20021031    500
           ... 
20230831    502
20230930    501
20231031    501
20231130    501
20231231    501
Length: 259, dtype: int64
20020630    500
20020731    500
20020831    500
20020930    500
20021031    500
           ... 
20230831    502
20230930    501
20231031    501
20231130    501
20231231    501
Length: 259, dtype: int64
20020630    500
20020731    500
20020831    500
20020930    500
20021031    500
           ... 
20230831    502
20230930    501
20231031    501
20231130    501
20231231    501
Length: 259, dtype: int64


Q5.

Correct the temporary reduction in Index members by doing the
following:
1.   Check for the dates where number of members are below 500
2.   Look at the number of members on the next available date
3.   If the number of members on the next date is higher, replace the values of the date with the shortfall with the next month’s values

To check homogeinity and debug

In [ ]:
member_counts = updated_result_df.sum(axis=1)
member_counts1 = updated_result_df1.sum(axis=1)
member_counts2 = updated_result_df2.sum(axis=1)
member_counts3 = updated_result_df3.sum(axis=1)
rows_below_500 = member_counts[member_counts < 500].index
rows_below_500_1 = member_counts[member_counts < 500].index
rows_below_500_2 = member_counts1[member_counts1 < 500].index
rows_below_500_3 = member_counts2[member_counts2 < 500].index
print(f"Number of rows with fewer than 500 members: {len(rows_below_500)}")
print(f"Number of rows with fewer than 500 members: {(rows_below_500)}")
print(f"Number of rows with fewer than 500 members: {len(rows_below_500_1)}")
print(f"Number of rows with fewer than 500 members: {(rows_below_500_1)}")
print(f"Number of rows with fewer than 500 members: {len(rows_below_500_2)}")
print(f"Number of rows with fewer than 500 members: {(rows_below_500_2)}")
print(f"Number of rows with fewer than 500 members: {len(rows_below_500_3)}")
print(f"Number of rows with fewer than 500 members: {(rows_below_500_3)}")


Number of rows with fewer than 500 members: 73
Number of rows with fewer than 500 members: Index(['20030930', '20031031', '20031130', '20031231', '20040131', '20040229',
       '20040331', '20040430', '20040531', '20040630', '20040731', '20040831',
       '20040930', '20041231', '20050131', '20050831', '20050930', '20051031',
       '20051130', '20051231', '20060331', '20060430', '20060531', '20060731',
       '20060831', '20071031', '20080331', '20110228', '20110331', '20110430',
       '20110531', '20110630', '20110731', '20110831', '20110930', '20111031',
       '20111130', '20111231', '20120131', '20120229', '20120331', '20120430',
       '20120531', '20120630', '20120731', '20120831', '20120930', '20121031',
       '20121130', '20121231', '20130131', '20130228', '20130331', '20130430',
       '20130531', '20130630', '20130731', '20130831', '20131130', '20140531',
       '20140831', '20141130', '20150131', '20151031', '20160131', '20160430',
       '20170731', '20170930', '20180331

Ans.\
Thought Process:
Identifying the Problem: The key problem is identifying dates with insufficient members and correcting them by copying values from the members on the next date that meets the required member count.
The approach uses a sequential search (starting from the deficient date) to find the next date with sufficient members(members>=500).
In our code if the next date also contains deficient members we leave that date as it is and move on. If the next date has sufficient memebers we fill the date with values of the next date.

In [ ]:
updated_result_df4=updated_result_df3.copy()

In [ ]:
min_required_members = 500
for date in rows_below_500_3:
    current_index = updated_result_df4.index.get_loc(date)
    print(current_index)
    print(date)
    print(current_index+1)
    print(len(updated_result_df4.index))
    while current_index + 1 < len(updated_result_df4.index):
        next_date_index = current_index + 1
        print(next_date_index)
        next_date = updated_result_df4.index[next_date_index]
        print(next_date)
        next_date_count = member_counts[next_date]
        print(next_date_count)
        if next_date_count>= min_required_members:
            updated_result_df4.loc[date] = updated_result_df4.loc[next_date]
            print(updated_result_df4.loc[date])
        elif next_date_count<= min_required_members:
          updated_result_df4.loc[date]=updated_result_df4.loc[date]
          break
        current_index = next_date_index
        date=next_date
        print(date)
        print(current_index)

print(updated_result_df4)

15
20030930
16
259
16
20031031
498
16
20031031
17
259
17
20031130
497
17
20031130
18
259
18
20031231
497
18
20031231
19
259
19
20040131
497
19
20040131
20
259
20
20040229
497
20
20040229
21
259
21
20040331
496
21
20040331
22
259
22
20040430
496
22
20040430
23
259
23
20040531
497
23
20040531
24
259
24
20040630
497
24
20040630
25
259
25
20040731
497
25
20040731
26
259
26
20040831
497
26
20040831
27
259
27
20040930
497
27
20040930
28
259
28
20041031
500
Stock
1909315D    1.0
1958850D    1.0
ABAN        1.0
3M          1.0
ABCAP       NaN
           ... 
ZNDU        1.0
ZUARIIND    1.0
ZEEN        NaN
ZCVCS       NaN
ZOMATO      NaN
Name: 20040930, Length: 1215, dtype: float64
20041031
28
29
20041130
500
Stock
1909315D    1.0
1958850D    1.0
ABAN        1.0
3M          1.0
ABCAP       NaN
           ... 
ZNDU        1.0
ZUARIIND    1.0
ZEEN        NaN
ZCVCS       NaN
ZOMATO      NaN
Name: 20041031, Length: 1215, dtype: float64
20041130
29
30
20041231
499
30
20041231
31
259
31
20050131
499


Q6.


1. The data currently is at monthly intervals.
2.  Expand the data to incorporate all the in-between dates
3. Fill the daily data with the latest available values as on that date.


Ans.
1.   We first generate a date range with daily frequency, covering the entire
period from the earliest to the latest date in updated_result_df4 using pd.date_range with frequency 'D'(daily).

2. We create a new empty DataFrame daily_df with the daily date range as its index

3.   index of daily_df is reformatted from the standard date format to YYYYMMDD (e.g., 20230831), to matche the format used in updated_result_df4.
4.   We then use left join to combine the two dataframes such that the new combined dataframe has the inbetween daily dates and not only the month end dates







In [ ]:
daily_index = pd.date_range(start=updated_result_df4.index.min(), end=updated_result_df4.index.max(), freq='D')
daily_df = pd.DataFrame(index=daily_index)
daily_df.index = daily_df.index.strftime('%Y%m%d')
combined_df = daily_df.join(updated_result_df4,how='left')
print(combined_df)

          1909315D  1958850D  ABAN   3M  ABCAP  360ONE  1918689D  294908Q  \
20020630       1.0       NaN   1.0  1.0    NaN     NaN       1.0      NaN   
20020701       NaN       NaN   NaN  NaN    NaN     NaN       NaN      NaN   
20020702       NaN       NaN   NaN  NaN    NaN     NaN       NaN      NaN   
20020703       NaN       NaN   NaN  NaN    NaN     NaN       NaN      NaN   
20020704       NaN       NaN   NaN  NaN    NaN     NaN       NaN      NaN   
...            ...       ...   ...  ...    ...     ...       ...      ...   
20231227       NaN       NaN   NaN  NaN    NaN     NaN       NaN      NaN   
20231228       NaN       NaN   NaN  NaN    NaN     NaN       NaN      NaN   
20231229       NaN       NaN   NaN  NaN    NaN     NaN       NaN      NaN   
20231230       NaN       NaN   NaN  NaN    NaN     NaN       NaN      NaN   
20231231       NaN       NaN   NaN  1.0    1.0     1.0       NaN      NaN   

          2299371D  20ML  ...  WESTLIFE  ZYDUSLIF  WDJL  ZC  ZDC  ZNDU  \
2

In [ ]:
# Count the number of rows where at least one column contains a '1'
rows_with_1 = (combined_df == 1).any(axis=1).sum()

print(f"Number of rows with at least one '1': {rows_with_1}")

Number of rows with at least one '1': 259


We use forward fill method to fill the inbetween data. The ffill() method replaces the NAN values with the value from the previous row.

In [ ]:
combined_df_ffill = combined_df.ffill()
print(combined_df_ffill)

          1909315D  1958850D  ABAN   3M  ABCAP  360ONE  1918689D  294908Q  \
20020630       1.0       NaN   1.0  1.0    NaN     NaN       1.0      NaN   
20020701       1.0       NaN   1.0  1.0    NaN     NaN       1.0      NaN   
20020702       1.0       NaN   1.0  1.0    NaN     NaN       1.0      NaN   
20020703       1.0       NaN   1.0  1.0    NaN     NaN       1.0      NaN   
20020704       1.0       NaN   1.0  1.0    NaN     NaN       1.0      NaN   
...            ...       ...   ...  ...    ...     ...       ...      ...   
20231227       1.0       1.0   1.0  1.0    1.0     1.0       1.0      1.0   
20231228       1.0       1.0   1.0  1.0    1.0     1.0       1.0      1.0   
20231229       1.0       1.0   1.0  1.0    1.0     1.0       1.0      1.0   
20231230       1.0       1.0   1.0  1.0    1.0     1.0       1.0      1.0   
20231231       1.0       1.0   1.0  1.0    1.0     1.0       1.0      1.0   

          2299371D  20ML  ...  WESTLIFE  ZYDUSLIF  WDJL   ZC  ZDC  ZNDU  \


In [ ]:
# Count the number of rows where at least one column contains a '1'
rows_with_1 = (combined_df_ffill == 1).any(axis=1).sum()

print(f"Number of rows with at least one '1': {rows_with_1}")

Number of rows with at least one '1': 7855


Question 7\
• Using Newton’s method, write a function to compute the square root of any number.\
Question 8\
• Can you think of a faster method to solve the above? If yes, please show

Question 7 and 8 has been doen together for ease


Ans.\
Mathematical Overview


---


Newton's Method for Square Roots:

Newton's method is an iterative technique for finding successively better approximations of the roots (or zeroes) of a real-valued function.
This method converges quadratically, meaning that the number of correct digits roughly doubles with each iteration.


---


Halley's Method for Square Roots:

Halley's method is a third-order iterative method for finding roots of functions. It can be considered an extension of Newton's method, using additional derivative information to achieve faster convergence.

This method converges cubically, meaning the number of correct digits triples with each iteration, leading to potentially faster convergence than Newton's method, especially when very high precision is required.


---


Karatsuba Algorithm for Square Roots:

The Karatsuba algorithm is primarily a fast multiplication method used for large integers. When adapted for square roots, it provides a way to handle large numbers efficiently by breaking the problem into smaller parts recursively.
The karatsuba_sqrt function essentially uses a method similar to Newton's method but relies on integer division and multiplication, which can be optimized for very large numbers due to the Karatsuba multiplication technique.


---


Performance Comparison

Halley's Method:

Advantage: Halley's method often has better execution time when high precision is required because it converges more quickly (cubic convergence) compared to Newton's method (quadratic convergence). This means fewer iterations are needed to reach a certain level of accuracy, making it faster in scenarios where precision is critical.
Disadvantage: The complexity of each iteration in Halley's method is higher due to the use of additional derivative calculations, which might make it slower if fewer iterations are needed or for problems requiring low precision.

Halley's Method excels in scenarios requiring high precision due to its faster convergence.


---



Karatsuba Square Root:

Advantage: The Karatsuba algorithm is particularly efficient for large integers because it reduces the complexity of multiplication. When applied to square roots, it can handle very large numbers effectively, making it faster for cases where the input size is extremely large.
Disadvantage: For smaller numbers or when the overhead of recursive calls outweighs the benefits of faster multiplication, Karatsuba may not be as efficient as Halley's or even Newton's method.

Karatsuba Square Root performs better with very large numbers because it optimizes the multiplication steps involved in iterative root-finding methods.


---



In [ ]:
import math
# Newton's Method for Square Root
def sqrt_newton(a, tolerance=1e-10, max_iterations=1000):
    if a < 0:
        raise ValueError("Cannot compute square root of a negative number")
    if a == 0:
        return 0
    x = a
    for _ in range(max_iterations):
        x_next = 0.5 * (x + a / x)
        if abs(x_next - x) < tolerance:
            return x_next
        x = x_next
    raise ValueError("Newton's method did not converge")

def sqrt_halley(a, tolerance=1e-10, max_iterations=1000):
    """Compute the square root of a number using Halley's method."""
    if a < 0:
        raise ValueError("Cannot compute square root of a negative number")

    if a == 0:
        return 0

    # Initial guess for the square root
    x = a
    for _ in range(max_iterations):
        f_x = x**2 - a
        f_prime_x = 2 * x
        f_double_prime_x = 2

        # Update the guess using Halley's method formula
        denominator = 2 * (f_prime_x**2 - f_x * f_double_prime_x)
        if denominator == 0:
            raise ValueError("Denominator in Halley's method is zero")

        x_next = x - (2 * f_x * f_prime_x) / denominator

        if abs(x_next - x) < tolerance:
            return x_next

        x = x_next

    raise ValueError("Halley's method did not converge")
def karatsuba(x, y):
    """Karatsuba multiplication algorithm."""
    if x < 10 or y < 10:
        return x * y

    m = min(len(str(x)), len(str(y)))
    m2 = m // 2

    # Split the digit sequences
    x1, x0 = divmod(x, 10**m2)
    y1, y0 = divmod(y, 10**m2)

    # 3 recursive calls
    z2 = karatsuba(x1, y1)
    z0 = karatsuba(x0, y0)
    z1 = karatsuba(x1 + x0, y1 + y0) - z2 - z0

    return z2 * 10**(2*m2) + z1 * 10**m2 + z0
def karatsuba_sqrt(n):
    """Compute the integer square root of n using Karatsuba multiplication."""
    if n < 0:
        raise ValueError("Cannot compute square root of a negative number")
    if n == 0:
        return 0

    # Initial guess (can be optimized)
    x = n

    while True:
        x_next = (x + n // x) // 2
        if x_next >= x:
            return x
        x = x_next

# Number to compute the square root of
number = 1234567890123451234567895678901234567890123456789056781111111111111187999999999334567

# Timing function
def time_method(method, *args):
    start_time = time.time()
    method(*args)
    return time.time() - start_time

# Measure execution times
time_newton = time_method(sqrt_newton, number)
time_halley = time_method(sqrt_halley, number)
time_karatsuba = time_method(karatsuba_sqrt, number)


print(f"Newton's Method: {time_newton:.6f} seconds")
print(f"Halley's Method: {time_halley:.6f} seconds")
print(f"Karatsuba Square Root: {time_karatsuba:.6f} seconds")

Newton's Method: 0.000074 seconds
Halley's Method: 0.000016 seconds
Karatsuba Square Root: 0.000070 seconds
